src/silo/query_engine/actions/tuple.cpp

```diff
@@ -0,0 +1,428 @@
+      const silo::Idx value2 = *reinterpret_cast<const silo::Idx*>(*data_pointer2);
+      *data_pointer2 += sizeof(decltype(value2));
+      const std::string string_value2 =
+         columns.pango_lineage_columns.at(metadata.name).lookupValue(value2).value;
+      return compareString(string_value1, string_value2);
+   }
+   if (metadata.type == ColumnType::INDEXED_STRING) {
+      const silo::Idx value1 = *reinterpret_cast<const silo::Idx*>(*data_pointer1);
+      *data_pointer1 += sizeof(decltype(value1));
+      const std::string string_value1 =
+         columns.indexed_string_columns.at(metadata.name).lookupValue(value1);
+      const silo::Idx value2 = *reinterpret_cast<const silo::Idx*>(*data_pointer2);
+      *data_pointer2 += sizeof(decltype(value2));
+      const std::string string_value2 =
+         columns.indexed_string_columns.at(metadata.name).lookupValue(value2);
+      return compareString(string_value1, string_value2);
+   }
+   if (metadata.type == ColumnType::INSERTION) {
+      const silo::Idx value1 = *reinterpret_cast<const silo::Idx*>(*data_pointer1);
+      *data_pointer1 += sizeof(decltype(value1));
+      const std::string string_value1 =
+         columns.insertion_columns.at(metadata.name).lookupValue(value1);
+      const silo::Idx value2 = *reinterpret_cast<const silo::Idx*>(*data_pointer2);
+      *data_pointer2 += sizeof(decltype(value2));
+      const std::string string_value2 =
+         columns.insertion_columns.at(metadata.name).lookupValue(value2);
+      return compareString(string_value1, string_value2);
+   }
+   throw std::runtime_error("Unchecked column type of column " + metadata.name);
+}
+
+size_t getColumnSize(const silo::storage::ColumnMetadata& metadata) {
+   if (metadata.type == silo::config::ColumnType::STRING) {
+      return sizeof(silo::common::String<silo::common::STRING_SIZE>);
+   } else if (metadata.type == silo::config::ColumnType::FLOAT) {
+      return sizeof(double);
+   } else if (metadata.type == silo::config::ColumnType::INT) {
+      return sizeof(int32_t);
+   } else if (metadata.type == silo::config::ColumnType::DATE) {
+      return sizeof(silo::common::Date);
+   } else {
+      return sizeof(silo::Idx);
+   }
+}
+
+}  // namespace
+
+size_t silo::query_engine::actions::getTupleSize(
+   const std::vector<silo::storage::ColumnMetadata>& metadata_list
+) {
+   size_t size = 0;
+   for (const auto& metadata : metadata_list) {
+      size += getColumnSize(metadata);
+   }
+   return size;
+}
+
+Tuple::Tuple(const silo::storage::ColumnPartitionGroup* columns, std::byte* data, size_t data_size)
+    : columns(columns),
+      data(data),
+      data_size(data_size) {}
+
+Tuple::Tuple(Tuple&& other) noexcept
+    : columns(other.columns),
+      data(std::exchange(other.data, nullptr)),
+      data_size(other.data_size) {}
+
+Tuple& Tuple::operator=(const Tuple& other) {
+   if (this == &other) {
+      return *this;
+   }
+   assert(this->data_size == other.data_size);
```

Comment:
I'd like to keep it in, as it would imply a serious programming error, but we do not want *any* runtime overhead here. So this leaves us with an assert or a workaround that mimics an assert

Reply:
In order to make our own life in the future more pleasant I would for the following two additions:
- a test that shows this behavior (and the possible things one could have gotten wrong) In tests the assert should fire, doesnt it?
- an error message, describing has gone wrong (when I recall it correctly we did only get a message saying "aborted", which is not very helpful. I saw on https://en.cppreference.com/w/cpp/error/assert that one can add text to the assert. Maybe this helps.)

Reply:
I do not know what else to add as the line number and the assert that failed are logged to stderr anyways